src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexInterpreter.cs

```diff
@@ -28,9 +76,454 @@ public RegexInterpreter(RegexCode code, CultureInfo culture)
+            // See the comment in StartTimeoutWatch().
+
+            int currentMillis = Environment.TickCount;
+
+            if (currentMillis < _timeoutOccursAt)
+                return;
+
+            if (0 > _timeoutOccursAt && 0 < currentMillis)
+                return;
+
+#if DEBUG
+            if (_runregex.Debug)
+            {
+                Debug.WriteLine("");
+                Debug.WriteLine("RegEx match timeout occurred!");
+                Debug.WriteLine("Specified timeout:       " + TimeSpan.FromMilliseconds(_timeout).ToString());
+                Debug.WriteLine("Timeout check frequency: " + TimeoutCheckFrequency);
+                Debug.WriteLine("Search pattern:          " + _runregex.pattern);
+                Debug.WriteLine("Input:                   " + input.ToString());
+                Debug.WriteLine("About to throw RegexMatchTimeoutException.");
+            }
+#endif
+
+            throw new RegexMatchTimeoutException(input.ToString(), _runregex.pattern, TimeSpan.FromMilliseconds(_timeout));
+        }
+
+        /// <summary>
+        /// Initializes all the data members that are used by Go()
+        /// </summary>
+        private void InitMatch(MemoryOrPinnedSpan<char> input)
+        {
+            // Use a hashtabled Match object if the capture numbers are sparse
+
+            if (_runmatch == null)
+            {
+                if (_runregex.caps != null)
+                    _runmatch = new MatchSparse(_runregex, _runregex.caps, _runregex.capsize, input, _runtextbeg, _runtextend - _runtextbeg, _runtextstart);
+                else
+                    _runmatch = new Match(_runregex, _runregex.capsize, input, _runtextbeg, _runtextend - _runtextbeg, _runtextstart);
+            }
+            else
+            {
+                _runmatch.Reset(_runregex, input, _runtextbeg, _runtextend, _runtextstart);
+            }
+
+            // note we test runcrawl, because it is the last one to be allocated
+            // If there is an alloc failure in the middle of the three allocations,
+            // we may still return to reuse this instance, and we want to behave
+            // as if the allocations didn't occur. (we used to test _trackcount != 0)
+
+            if (_runcrawl != null)
+            {
+                _runtrackpos = _runtrack.Length;
+                _runstackpos = _runstack.Length;
+                _runcrawlpos = _runcrawl.Length;
+                return;
+            }
+
+            InitTrackCount();
+
+            int tracksize = _runtrackcount * 8;
+            int stacksize = _runtrackcount * 8;
+
+            if (tracksize < 32)
+                tracksize = 32;
+            if (stacksize < 16)
+                stacksize = 16;
+
+            _runtrack = new int[tracksize];
+            _runtrackpos = tracksize;
+
+            _runstack = new int[stacksize];
+            _runstackpos = stacksize;
+
+            _runcrawl = new int[32];
```

Comment:
I'm only skimming this code as I'm assuming it's mainly the same as what's already there, but, there are lots of array allocations here and throughout.  Can one or more be done with ArrayPool or ValueListBuilder or some such thing?

Reply:
You are right, there are plenty of places in RegexInterpreter that could be improved allocation-wise. Maybe the whole interpreter could be changed to a ref struct similar to what I did in other places in Regex in combination with a factory which holds the already per-computed values to construct the interpreter. I have several ideas in my head that I first need to think through.

Reply:
... not really necessary for the design review but later for the perf measurements (as I noticed some regression during my first benchmarks).